src/plot_api/manage_arrays.js

```diff
@@ -0,0 +1,211 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var nestedProperty = require('../lib/nested_property');
+var isPlainObject = require('../lib/is_plain_object');
+var noop = require('../lib/noop');
+var Loggers = require('../lib/loggers');
+var Registry = require('../registry');
+
+
+exports.containerArrayMatch = require('./container_array_match');
+
+var isAddVal = exports.isAddVal = function isAddVal(val) {
+    return val === 'add' || isPlainObject(val);
+};
+
+var isRemoveVal = exports.isRemoveVal = function isRemoveVal(val) {
+    return val === null || val === 'remove';
+};
+
+/*
+ * editContainerArray: for managing arrays of layout components in relayout
+ * handles them all with a consistent interface.
+ *
+ * Here are the supported actions -> relayout calls -> edits we get here
+ * (as prepared in _relayout):
+ *
+ * add an empty obj -> {'annotations[2]': 'add'} -> {2: {'': 'add'}}
+ * add a specific obj -> {'annotations[2]': {attrs}} -> {2: {'': {attrs}}}
```

Comment:
right - it's a bit of a weird structure I ended up with as input to `editContainerArray`. It's probably clear, but this is prepared in [`_relayout`](https://github.com/plotly/plotly.js/pull/1403/files#diff-2941ab69a12080c0633ff4ac8ea3aa83R2024) and users never see it. It's structured as `edits[index][attribute]` - so if you're setting an array element all at once, `attribute` will be `''`, and if you're setting the whole array at once, both `attribute` and `index` will be `''`